src/Test/Expectation.elm

```diff
@@ -1,16 +1,39 @@
-module Test.Expectation exposing (Expectation(..), withGiven)
+module Test.Expectation exposing (Expectation(..), Reason(..), withGiven)
 
 
 type Expectation
     = Pass
-    | Fail String String
+    | Fail { given : String, description : String, reason : Reason }
 
 
+
+-- TODO: given : Maybe String
+
+
+type Reason
+    = Custom
+    | Equals String String
+    | Comparison String String
+      -- Expected, actual, (index of problem, expected element, actual element)
+    | ListDiff String String ( Int, String, String )
+      {- I don't think we need to show the diff twice with + and - reversed. Just show it after the main vertical bar.
+         "Extra" and "missing" are relative to the actual value.
+      -}
+    | CollectionDiff
+        { expected : String
+        , actual : String
+        , extra : List String
+        , missing : List String
+        }
```

Comment:
No need for it to block this PR, but just a quick note that I already want to add a branch here for https://github.com/eeue56/elm-html-test semantic errors. It'll probably look something like:

```elm
| HtmlDiff
    { reason : Reason
    , queries : List { query : String, html : List String }
    }
```
I want to do that in a later PR though. Need to figure out what the other side needs first.